### PR TITLE
[vcpkg] Refactor toolchain & generator selection

### PR DESCRIPTION
**Pull request by Neumann-A**
picked from #23259

### CODE REVIEW

**Comment by Neumann-A** [comment]
bento4:arm-uwp regressions: 
`    30>CUSTOMBUILD : CMake error : Cannot restore timestamp "D:/buildtrees/bento4/arm-uwp-dbg/CMakeFiles/generate.stamp": Access is denied. [D:\buildtrees\bento4\arm-uwp-dbg\mp4dump.vcxproj]`
Maybe a simple retry fixes it.

**Comment by dg0yt** [comment]
> bento4:arm-uwp regressions: ` 30>CUSTOMBUILD : CMake error : Cannot restore timestamp "D:/buildtrees/bento4/arm-uwp-dbg/CMakeFiles/generate.stamp": Access is denied. [D:\buildtrees\bento4\arm-uwp-dbg\mp4dump.vcxproj]` Maybe a simple retry fixes it.

I had a similar issue with nanoflann: https://github.com/microsoft/vcpkg/pull/21507#issuecomment-1079643706
I found that CMake has undocumented options for retries and waiting, but I don't have the reference now. (It is in Kitware's tickets, related to network file systems.)

File: ports/vcpkg-cmake/vcpkg_cmake_configure.cmake [hunk 1]
```diff
@@ -159,71 +159,18 @@ function(vcpkg_cmake_configure)
...
-        endif()
-    else()
-        set(generator "Ninja")
+        set(generator "")
+        set(arch "")
+        z_vcpkg_get_visual_studio_generator(generator arch)
     endif()
 
-    if(NOT DEFINED generator)
+    if(NOT DEFINED generator  OR generator STREQUAL "")
```
Comment (Thomas1664) [inline ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:171]:
```suggestion
    if(NOT DEFINED generator OR generator STREQUAL "")
```

Can you remove the double whitespace?

Comment (dg0yt) [reply ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:171]: IMO it should be just `IF(NOT generator)`. But there is a different change in #21507 anyways.

Comment (Neumann-A) [reply ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:171]: I don't feel like forcing CI to do a big rebuild due to just a WS change. Especially if the section will be changed by other PRs anyway.

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
I'd like @strega-nil-ms review this PR again.

File: scripts/cmake/vcpkg_configure_cmake.cmake [hunk 2]
```diff
@@ -159,38 +248,13 @@ function(vcpkg_configure_cmake)
...
-
-        if("${VCPKG_TARGET_ARCHITECTURE}" STREQUAL "x86")
-            set(generator_arch "Win32")
-        elseif("${VCPKG_TARGET_ARCHITECTURE}" STREQUAL "x64")
-            set(generator_arch "x64")
-        elseif("${VCPKG_TARGET_ARCHITECTURE}" STREQUAL "arm")
-            set(generator_arch "ARM")
-        elseif("${VCPKG_TARGET_ARCHITECTURE}" STREQUAL "arm64")
-            set(generator_arch "ARM64")
-        endif()
```
Comment (strega-nil) [inline scripts/cmake/vcpkg_configure_cmake.cmake:192]: Can you use this as the basis for `z_vcpkg_get_visual_studio_generator` instead?

File: scripts/cmake/vcpkg_configure_cmake.cmake [hunk 3]
```diff
@@ -98,6 +98,91 @@ function(z_vcpkg_configure_cmake_build_cmakecache out_var whereat build_type)
     set("${out_var}" "${${out_var}}${line}\"\n\n" PARENT_SCOPE)
 endfunction()
 
+function(z_vcpkg_get_visual_studio_generator generator_var arch_var)
+    set(arch "")
```
Comment (strega-nil) [inline scripts/cmake/vcpkg_configure_cmake.cmake:102]:
```suggestion
function(z_vcpkg_get_visual_studio_generator)
    cmake_parse_arguments(PARSE_ARGV 0 arg "" "OUT_GENERATOR;OUT_ARCH" "")
    
    if (NOT DEFINED arg_OUT_GENERATOR)
        message(FATAL_ERROR "OUT_GENERATOR must be defined.")
    endif()
    if(NOT DEFINED arg_OUT_ARCH)
        message(FATAL_ERROR "OUT_ARCH must be defined.")
    endif()
    if(DEFINED arg_UNPARSED_ARGUMENTS)
            message(FATAL_ERROR "${CMAKE_CURRENT_FUNCTION} was passed extra arguments: ${arg_UNPARSED_ARGUMENTS}")
    endif()
```

**Review by strega-nil — CHANGES_REQUESTED** [review]
Those two changes then I'm an approve

File: scripts/cmake/vcpkg_configure_cmake.cmake (hunk 2 above)
Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_cmake.cmake:192]: done

**Comment by JackBoosY** [comment]
baseline issues.

**Comment by JackBoosY** [comment]
Depends on https://github.com/microsoft/vcpkg/pull/23999 and https://github.com/microsoft/vcpkg/pull/23945.

**Review by JackBoosY — APPROVED** [review]

**Review by strega-nil-ms — APPROVED** [review]

**Comment by strega-nil-ms** [comment]
Thanks @Neumann-A !

**Pull request merged**